File: app/src/aether/tools/schedule_reminder.py
"""ScheduleReminderTool — let the LLM schedule future user reminders.

The LLM calls this tool when the user asks to be reminded of something at a
future time.  The tool posts a one-shot cron job to the orchestrator; when
the job fires, the orchestrator delivers a /cron_event back to the agent and
the LLM relays the reminder to the user in natural language.

Example LLM call:
    schedule_reminder(
        message="Remind the user their standup meeting starts in 5 minutes.",
        iso_datetime="2026-02-21T09:55:00Z",
    )
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone

from aether.kernel.cron import CronClient
from aether.tools.base import AetherTool, ToolParam, ToolResult

logger = logging.getLogger(__name__)


class ScheduleReminderTool(AetherTool):
    """Schedule a one-shot reminder to be delivered to the user at a future time."""

    name = "schedule_reminder"
    description = (
        "Schedule a reminder to be delivered to the user at a specific future date and time. "
        "Use this when the user asks to be reminded of something later. "
        "The reminder will be spoken or shown to the user at the requested time."
    )
    parameters = [
        ToolParam(
            name="message",
            type="string",
            description=(
                "The reminder message to deliver to the user. "
                "Write it as a natural spoken sentence, e.g. "
                "'Remind the user their dentist appointment is in 10 minutes.'"
            ),
            required=True,
        ),
        ToolParam(
            name="iso_datetime",
            type="string",
            description=(
                "The exact date and time to deliver the reminder, in ISO-8601 format "
                "(e.g. '2026-02-21T15:30:00Z'). Must be in the future."
            ),
            required=True,
        ),
    ]

    def __init__(self) -> None:
        self._cron = CronClient()

    async def execute(self, message: str, iso_datetime: str) -> ToolResult:  # type: ignore[override]
        """Schedule the reminder via the orchestrator cron API."""
        # Parse and validate the datetime
        try:
            run_at = datetime.fromisoformat(iso_datetime.replace("Z", "+00:00"))
        except ValueError:
            return ToolResult.fail(
                f"Invalid datetime format: {iso_datetime!r}. Use ISO-8601, e.g. '2026-02-21T15:30:00Z'."
            )

        if run_at.tzinfo is None:
            run_at = run_at.replace(tzinfo=timezone.utc)

        now = datetime.now(tz=timezone.utc)
        if run_at <= now:
            return ToolResult.fail(
                f"The requested time {iso_datetime} is in the past. Please provide a future time."
            )

        # Build the instruction the LLM will receive when the cron fires
        instruction = (
            f"Scheduled reminder: {message} "
            f"Please relay this reminder to the user now in a natural, friendly way."
        )

        job_id = await self._cron.schedule(
            instruction=instruction,
            run_at=run_at,
            plugin=None,  # not plugin-specific
            interval_s=None,  # one-shot
        )

        if not job_id:
            return ToolResult.fail(
                "Failed to schedule the reminder — the orchestrator could not be reached. Please try again."
            )

        friendly_time = run_at.strftime("%B %d, %Y at %H:%M UTC")
        logger.info(
            "Reminder scheduled: job_id=%s, run_at=%s", job_id, run_at.isoformat()
        )

        return ToolResult.success(
            f"Reminder scheduled for {friendly_time}. I'll remind you then.",
            job_id=job_id,
            run_at=run_at.isoformat(),
        )

        if run_at.tzinfo is None:
            run_at = run_at.replace(tzinfo=timezone.utc)

        now = datetime.now(tz=timezone.utc)
        if run_at <= now:
            return ToolResult(
                success=False,
                output=f"The requested time {iso_datetime} is in the past. Please provide a future time.",
            )

        # Build the instruction the LLM will receive when the cron fires
        instruction = (
            f"Scheduled reminder: {message} "
            f"Please relay this reminder to the user now in a natural, friendly way."
        )

        job_id = await self._cron.schedule(
            instruction=instruction,
            run_at=run_at,
            plugin=None,  # not plugin-specific
            interval_s=None,  # one-shot
        )

        if not job_id:
            return ToolResult(
                success=False,
                output="Failed to schedule the reminder — the orchestrator could not be reached. Please try again.",
            )

        friendly_time = run_at.strftime("%B %d, %Y at %H:%M UTC")
        logger.info(
            "Reminder scheduled: job_id=%s, run_at=%s", job_id, run_at.isoformat()
        )

        return ToolResult(
            success=True,
            output=f"Reminder scheduled for {friendly_time}. I'll remind you then.",
            metadata={"job_id": job_id, "run_at": run_at.isoformat()},
        )

File: agent/internal/tools/builtin/schedule_reminder.go
package builtin

import (
	"context"
	"fmt"
	"time"

	"github.com/suryaumapathy/core-ai/agent/internal/reminders"
	"github.com/suryaumapathy/core-ai/agent/internal/tools"
)

type ScheduleReminderTool struct{}

func (t *ScheduleReminderTool) Definition() tools.Definition {
	return tools.Definition{
		Name:        "schedule_reminder",
		Description: "Schedule a one-shot reminder at a specific future time.",
		StatusText:  "Scheduling reminder...",
		Parameters: []tools.Param{
			{Name: "message", Type: "string", Description: "Reminder message", Required: true},
			{Name: "iso_datetime", Type: "string", Description: "Future datetime in RFC3339/ISO-8601", Required: true},
		},
	}
}

func (t *ScheduleReminderTool) Execute(ctx context.Context, call tools.Call) tools.Result {
	if call.Ctx.Store == nil {
		return tools.Fail("Scheduler store is unavailable", nil)
	}
	message, _ := call.Args["message"].(string)
	raw, _ := call.Args["iso_datetime"].(string)
	runAt, err := parseISOTime(raw)
	if err != nil {
		return tools.Fail("Invalid datetime format: "+raw+". Use ISO-8601, e.g. 2026-02-21T15:30:00Z", nil)
	}
	if !runAt.After(time.Now().UTC()) {
		return tools.Fail("The requested time is in the past. Please provide a future time.", nil)
	}
	scope := call.Ctx.Store.ScopeCronModule(reminders.CronModuleReminders)
	job, err := scope.ScheduleOnce(ctx, reminders.CronJobTypeDeliver, map[string]any{"message": message}, runAt, 5)
	if err != nil {
		return tools.Fail("Failed to schedule reminder: "+err.Error(), nil)
	}
	return tools.Success(
		fmt.Sprintf("Reminder scheduled for %s.", runAt.Format("January 02, 2006 at 15:04 MST")),
		map[string]any{"job_id": job.ID, "run_at": runAt.Format(time.RFC3339), "module": reminders.CronModuleReminders},
	)
}

func parseISOTime(v string) (time.Time, error) {
	if t, err := time.Parse(time.RFC3339, v); err == nil {
		return t.UTC(), nil
	}
	return time.Time{}, fmt.Errorf("invalid datetime")
}

var _ tools.Tool = (*ScheduleReminderTool)(nil)
